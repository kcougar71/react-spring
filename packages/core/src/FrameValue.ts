import {
  deprecateInterpolate,
  each,
  FluidValue,
  FluidObserver,
  Globals as G,
} from '@react-spring/shared'
import { InterpolatorArgs } from '@react-spring/types'
import { getAnimated } from '@react-spring/animated'

import { Interpolation } from './Interpolation'

export const isFrameValue = (value: any): value is FrameValue =>
  value instanceof FrameValue

let nextId = 1

/**
 * A kind of `FluidValue` that manages an `AnimatedValue` node.
 *
 * Its underlying value can be accessed and even observed.
 */
export abstract class FrameValue<T = any>
  extends FluidValue<T, FrameValue.Event<T>>
  implements FluidObserver<FrameValue.Event> {
  readonly id = nextId++

  abstract key?: string
  abstract get idle(): boolean

  protected _priority = 0
  protected _children = new Set<FrameValue.Observer<T>>()

  get priority() {
    return this._priority
  }
  set priority(priority: number) {
    if (this._priority != priority) {
      this._priority = priority
      this._onPriorityChange(priority)
    }
  }

  /** Get the current value */
  get(): T {
    const node = getAnimated(this)
    return node && node.getValue()
  }

  /** Create a spring that maps our value to another value */
  to<Out>(...args: InterpolatorArgs<T, Out>) {
    return G.to(this, args) as Interpolation<T, Out>
  }

  /** @deprecated Use the `to` method instead. */
  interpolate<Out>(...args: InterpolatorArgs<T, Out>) {
    deprecateInterpolate()
    return G.to(this, args) as Interpolation<T, Out>
  }

  toJSON() {
    return this.get()
  }

  /** @internal */
  addChild(child: FrameValue.Observer<T>): void {
    if (!this._children.size) this._attach()
    this._children.add(child)
  }

  /** @internal */
  removeChild(child: FrameValue.Observer<T>): void {
    this._children.delete(child)
    if (!this._children.size) this._detach()
  }

  /** @internal */
  abstract advance(dt: number): void

  /** @internal */
  abstract onParentChange(_event: FrameValue.Event): void

  /** Called when the first child is added. */
  protected _attach() {}

  /** Called when the last child is removed. */
  protected _detach() {}

  /** Tell our children about our new value */
  protected _onChange(value: T, idle = false) {
    this._emit({
      type: 'change',
      parent: this,
      value,
      idle,
    })
  }

  /** Tell our children about our new priority */
  protected _onPriorityChange(priority: number) {
    if (!this.idle) {
      // Make the frameloop aware of our new priority.
      G.frameLoop.start(this)
    }
    this._emit({
      type: 'priority',
      parent: this,
      priority,
    })
  }

  protected _emit(event: FrameValue.Event) {
    // Clone "_children" so it can be safely mutated inside the loop.
    each(Array.from(this._children), child => {
      child.onParentChange(event)
    })
  }
}

export declare namespace FrameValue {
  /** A parent changed its value */
  interface ChangeEvent<T = any> {
    parent: FrameValue<T>
    type: 'change'
    value: T
    idle: boolean
  }

  /** A parent changed its priority */
  interface PriorityEvent<T = any> {
    parent: FrameValue<T>
    type: 'priority'
    priority: number
  }

  /** A parent reset the internal state of its current animation */
  interface ResetEvent<T = any> {
    parent: FrameValue<T>
    type: 'reset'
  }

  /** A parent is done animating */
  interface IdleEvent<T = any> {
    parent: FrameValue<T>
    type: 'idle'
  }

  /** Events sent to children of `FrameValue` objects */
  export type Event<T = any> =
    | ChangeEvent<T>
    | PriorityEvent<T>
    | ResetEvent<T>
    | IdleEvent<T>

  /** An object that handles `FrameValue` events */
  export type Observer<T = any> = FluidObserver<Event<T>>
}
