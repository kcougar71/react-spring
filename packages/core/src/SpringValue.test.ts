import { SpringValue } from './SpringValue'
import { FrameValue } from './FrameValue'

const frameLength = 1000 / 60

describe('SpringValue', () => {
  it('can animate from 0 to 1 with linear easing', () => {
    const spring = new SpringValue(0)
    spring.start(1, {
      config: { duration: 10 * frameLength },
    })
    await advanceUntilIdle()
    const frames = getFrames(spring)
    expect(frames).toMatchSnapshot()
  })

  describeProps()

  describe('"set" method', () => {
    it('stops the active animation', async () => {
      const spring = new SpringValue(0)
      const promise = spring.start(1)

      await advanceUntilValue(spring, 0.5)
      spring.set(2)

      expect(spring.idle).toBeTruthy()
      expect(await promise).toMatchObject({
        finished: false,
        value: 2,
      })
    })

    describe('when a new value is passed', () => {
      it('calls the "onChange" prop', () => {
        const onChange = jest.fn()
        const spring = new SpringValue(0, { onChange })
        spring.set(1)
        expect(onChange).toBeCalledWith(1, spring)
      })
      it.todo('wraps the "onChange" call with "batchedUpdates"')
    })

    describe('when the current value is passed', () => {
      it('skips the "onChange" call', () => {
        const onChange = jest.fn()
        const spring = new SpringValue(0, { onChange })
        spring.set(0)
        expect(onChange).not.toBeCalled()
      })
    })
  })

  describeTarget('another SpringValue', from => {
    const node = new SpringValue(from)
    return {
      node,
      set: node.set.bind(node),
      start: node.start.bind(node),
      reset: node.reset.bind(node),
    }
  })

  describeTarget('an Interpolation', from => {
    const parent = new SpringValue(from - 1)
    const node = parent.to(n => n + 1)
    return {
      node,
      set: n => parent.set(n - 1),
      start: n => parent.start(n - 1),
      reset: parent.reset.bind(parent),
    }
  })
})

function describeProps() {
  describe('when "from" prop is changed', () => {
    describe('before the first animation', () => {
      it('updates the current value', () => {})
    })
    describe('after the first animation', () => {
      it.todo('does nothing unless "reset" is true')
    })
  })

  describe('when "to" prop is changed', () => {
    it.todo('resolves the "start" promise with (finished: false)')
    it.todo('avoids calling the "onStart" prop')
    it.todo('avoids calling the "onRest" prop')
  })

  describe('when "reset" prop is true', () => {
    it.todo('resolves the "start" promise with (finished: false)')
    it.todo('calls the "onRest" prop with (finished: false)')
  })

  describe('when "immediate" prop is true', () => {
    it.todo('still resolves the "start" promise')
    it.todo('never calls the "onStart" prop')
    it.todo('never calls the "onRest" prop')
  })
}

/** The minimum requirements for testing a dynamic target */
type OpaqueTarget = {
  node: FrameValue
  set: (value: number) => any
  start: (value: number) => Promise<any>
  reset: () => void
}

function describeTarget(name: string, create: (from: number) => OpaqueTarget) {
  let spring: SpringValue
  let target: OpaqueTarget
  beforeEach(() => {
    spring = new SpringValue(0)
    target = create(1)
  })

  describe('when our target is ' + name, () => {
    it('animates toward the current value', async () => {
      spring.start({ to: target.node })
      expect(spring.priority).toBeGreaterThan(target.node.priority)
      expect(spring.animation).toMatchObject({
        to: target.node,
        toValues: null,
      })

      await advanceUntilIdle()
      expect(spring.get()).toBe(target.node.get())
    })

    it.todo('preserves its "onRest" prop between animations')

    it('can change its target while animating', async () => {
      spring.start({ to: target.node })
      await advanceUntilValue(spring, target.node.get() / 2)

      spring.start(0)
      expect(spring.priority).toBe(0)
      expect(spring.animation).toMatchObject({
        to: 0,
        toValues: [0],
      })

      await advanceUntilIdle()
      expect(spring.get()).toBe(0)
    })

    describe('when target is done animating', () => {
      it('keeps animating until the target is reached', async () => {
        spring.start({ to: target.node })
        target.start(1.1)

        await advanceUntil(() => target.node.idle)
        expect(spring.idle).toBeFalsy()

        await advanceUntilIdle()
        expect(spring.idle).toBeTruthy()
        expect(spring.get()).toBe(target.node.get())
      })
    })

    describe('when target animates after we go idle', () => {
      it('starts animating', async () => {
        spring.start({ to: target.node })
        await advanceUntil(() => spring.idle)

        target.start(2)
        await advanceUntilIdle()

        expect(getFrames(spring).length).toBeGreaterThan(1)
        expect(spring.get()).toBe(target.node.get())
      })
    })

    describe('when target has its value set (not animated)', () => {
      it('animates toward the new value', async () => {
        spring.start({ to: target.node })
        await advanceUntilIdle()

        target.set(2)
        await advanceUntilIdle()

        expect(getFrames(spring).length).toBeGreaterThan(1)
        expect(spring.get()).toBe(target.node.get())
      })
    })

    describe('when target resets its animation', () => {
      it('keeps animating', async () => {
        spring.start({ to: target.node })
        target.start(2)

        await advanceUntilValue(target.node, 1.5)
        expect(target.node.idle).toBeFalsy()

        target.reset()
        expect(target.node.get()).toBe(1)

        await advanceUntilIdle()
        const frames = getFrames(spring)

        expect(frames.length).toBeGreaterThan(1)
        expect(spring.get()).toBe(target.node.get())
      })
    })

    describe('when animating a string', () => {
      it.todo('animates as expected')
    })

    describe('when animating an array', () => {
      it.todo('animates as expected')
    })
  })
}
