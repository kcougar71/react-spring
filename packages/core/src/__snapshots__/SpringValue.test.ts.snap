// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpringValue can animate a number 1`] = `
Array [
  0.09999999999999999,
  0.19999999999999998,
  0.3,
  0.39999999999999997,
  0.5,
  0.6,
  0.7000000000000001,
  0.7999999999999999,
  0.8999999999999999,
  0.9999999999999998,
  1,
]
`;

exports[`SpringValue can animate a string 1`] = `
Array [
  "0.9999999999999999px 1.9999999999999998px",
  "1.9999999999999998px 3.9999999999999996px",
  "3px 6px",
  "3.9999999999999996px 7.999999999999999px",
  "5px 10px",
  "6px 12px",
  "7.000000000000001px 14.000000000000002px",
  "7.999999999999999px 15.999999999999998px",
  "9px 18px",
  "9.999999999999998px 19.999999999999996px",
  "10px 20px",
]
`;

exports[`SpringValue can animate an array of numbers 1`] = `Array []`;

exports[`SpringValue when "reverse" prop is true swaps the "to" and "from" props 1`] = `
Array [
  0.9773651566921419,
  0.9238447110500605,
  0.852727045431224,
  0.7732066706218559,
  0.6914946808396315,
  0.6116475578389905,
  0.5361828794607453,
  0.4665342551963084,
  0.4033853476869672,
  0.3469132819682177,
  0.29696442635758724,
  0.2531799342583708,
  0.21508416627703766,
  0.18214585932574526,
  0.15381943659801983,
  0.1295719767229399,
  0.10889994138353111,
  0.09133869011601468,
  0.07646700587939456,
  0.06390825164952882,
  0.05332932801191841,
  0.0444382672440135,
  0.03698105225497496,
  0.030738067326888108,
  0.02552045540916056,
  0.02116656126728198,
  0.01753857162892271,
  0.014519415475703344,
  0.012009954436804516,
  0.009926470833638602,
  0.008198446305043062,
  0.006766614882944518,
  0.005581269243973591,
  0.0046007964253405855,
  0.0037904186834741348,
  0.003121115756173977,
  0.0025687061097373425,
  0.002113066492326755,
  0.0017374710529361444,
  0.0014280332726546588,
  0.001173235894326376,
  0,
]
`;

exports[`SpringValue when "reverse" prop is true works when "from" was set by an earlier update 1`] = `
Array [
  0.9773651566921419,
  0.9238447110500605,
  0.852727045431224,
  0.7732066706218559,
  0.6914946808396315,
  0.6116475578389905,
  0.5361828794607453,
  0.4665342551963084,
  0.4033853476869672,
  0.3469132819682177,
  0.29696442635758724,
  0.2531799342583708,
  0.21508416627703766,
  0.18214585932574526,
  0.15381943659801983,
  0.1295719767229399,
  0.10889994138353111,
  0.09133869011601468,
  0.07646700587939456,
  0.06390825164952882,
  0.05332932801191841,
  0.0444382672440135,
  0.03698105225497496,
  0.030738067326888108,
  0.02552045540916056,
  0.02116656126728198,
  0.01753857162892271,
  0.014519415475703344,
  0.012009954436804516,
  0.009926470833638602,
  0.008198446305043062,
  0.006766614882944518,
  0.005581269243973591,
  0.0046007964253405855,
  0.0037904186834741348,
  0.003121115756173977,
  0.0025687061097373425,
  0.002113066492326755,
  0.0017374710529361444,
  0.0014280332726546588,
  0.001173235894326376,
  0,
]
`;

exports[`SpringValue when "reverse" prop is true works when "to" and "from" were set by an earlier update 1`] = `
Array [
  0.022634843307857987,
  0.07615528894993949,
  0.14727295456877587,
  0.226793329378144,
  0.30850531916036833,
  0.3883524421610094,
  0.4638171205392546,
  0.5334657448036914,
  0.5739798090051748,
  0.5769314290818427,
  0.5557626190736369,
  0.5200267363634852,
  0.4764105145625939,
  0.4295016985132453,
  0.38236344286272556,
  0.3369622784733687,
  0.29448540630343617,
  0.25557459185220316,
  0.22049742047819276,
  0.18927168260884208,
  0.16175483826511933,
  0.1377075920817319,
  0.11683838434304501,
  0.0988339093960519,
  0.08337948597437063,
  0.07017212884873279,
  0.05892843425047188,
  0.049388836173825494,
  0.041319373575113914,
  0.03451179641037492,
  0.028782605949931907,
  0.023971452443943598,
  0.019939186165186966,
  0.016565764841941392,
  0.013748152945448568,
  0.011398299719529353,
  0.009441248327067157,
  0.007813404341311832,
  0.006460975252106907,
  0.005338581610289848,
  0.004408033349647206,
  0.00363726055737456,
  0.0029993856652006856,
  0.0024719230850121722,
  0.002036092265339304,
  0.0016762306751292224,
  0.0013792940991652848,
  0.001134432694629231,
  0.0009326324005005715,
  0.0007664124372832261,
  0.0006295707368567502,
  0,
]
`;
